helix-core/src/test/java/org/apache/helix/integration/TestAlertingRebalancerFailure.java

```diff
@@ -161,7 +170,8 @@ public void testParticipantUnavailable() throws Exception {
   @Test(dependsOnMethods = "testParticipantUnavailable")
   public void testTagSetIncorrect() throws Exception {
     _gSetupTool.addResourceToCluster(CLUSTER_NAME, testDb, 5,
-        BuiltInStateModelDefinitions.MasterSlave.name(), RebalanceMode.FULL_AUTO.name());
```

Comment:
It works in production. But it does not work in the test.
The reason is that our verifier does not read the controller cache (which is only used by the DelayedAutoRebalancer).
So the controller only does one calculation, and later pipeline will only read from the cache. But the verifier is calculating from scratch every time. And since the AutoRebalanceStrategy is not stable, the verify result will never converge.
If we change to CrushEd, then even the verifier computing from scratch, the result would be the same. And our tests can pass.

Reply:
The argument here is that we do want to deprecate AutoRebalancer and AutoRebalanceStrategy. Since they are buggy and unstable. So even the behavior change will hurt some of their use cases, we will take it.
If this assumption does not hold, then we can never change the default rebalancer to DelayedAutoRebalancer.

Reply:
So controller cache only works for DelayedAutoRebalancer and AutoRebalanceStrategy? I am not sure exactly where it is. Can you point to a code link?

Reply:
Synced offline with Kai. It is clear now.